### PR TITLE
Add tokenizer.json (HuggingFace fast tokenizer) support

### LINKED ISSUE #21: Add tokenizer.json (HuggingFace fast tokenizer) support to LoadTokenizer
## Summary

The `TokenizerPath` parameter does not support HuggingFace's `tokenizer.json` file — the single-file format that contains a complete tokenizer pipeline (vocabulary, merges, normalizer, pre/post-processing). This is the canonical format for HuggingFace "fast tokenizers" and is present on virtually every model on the Hub.

Adding support would let users point `TokenizerPath` at a `tokenizer.json` file (or a directory containing one) and have it "just work"  simplifying model packaging to a single tokenizer file instead of managing 2+ files.

Related: #18 (original feature request)
Depends on: #19 (adds DebertaV2Tokenizer + spm.model support  merge first)

---

## Background: What is `tokenizer.json`?

It's a **self-contained JSON file** produced by HuggingFace's Rust-based `tokenizers` library. It contains the ENTIRE tokenizer pipeline in one file:

```json
{
  "version": "1.0",
  "truncation": { ... },
  "padding": { ... },
  "added_tokens": [ { "id": 0, "content": "<s>", "special": true }, ... ],
  "normalizer": null,
  "pre_tokenizer": { "type": "ByteLevel", ... },
  "post_processor": { "type": "RobertaProcessing", ... },
  "decoder": { "type": "ByteLevel", ... },
  "model": {
    "type": "BPE",                              // <-- THIS is what we care about
    "vocab": { "<s>": 0, "Ġthe": 5, ... },     // full vocab as a JSON dict
    "merges": [ "Ġ t", "Ġ a", ... ]            // full merge rules as a JSON array
  }
}
```

The `model` section contains one of three types:
| `model.type` | Underlying algorithm | Example models |
|---|---|---|
| `"BPE"` | Byte Pair Encoding | RoBERTa, GPT-2, BART |
| `"WordPiece"` | WordPiece | BERT, DistilBERT, Electra |
| `"Unigram"` | SentencePiece Unigram | T5, XLNet, DeBERTa, Albert |

### Scope: Only BPE and WordPiece need implementation

**Unigram/SentencePiece models always ship a `.model` protobuf file alongside `tokenizer.json`** (the protobuf is produced by SentencePiece training; `tokenizer.json` is generated from it). The existing code already loads `.model` files via `LlamaTokenizer.Create()`. There is no supported way to convert a Unigram `tokenizer.json` back to a `.model` file, so we should NOT attempt it.

If someone points at a `tokenizer.json` with `model.type = "Unigram"`, the code should throw a clear error saying: "Unigram tokenizer.json is not supported  point TokenizerPath at the directory containing the .model file instead, or use the Tokenizer property."

---

## File to Modify

**`src/MLNet.TextInference.Onnx/TextTokenizerEstimator.cs`**  this is the ONLY file that needs changes.

---

### Step 1: Handle `tokenizer.json` as a direct file path in `LoadTokenizer()` (~line 163)

Currently, `LoadTokenizer()` checks if the path is a directory, then checks for `tokenizer_config.json`, then falls through to `LoadFromVocabFile()`. We need to add a check for `tokenizer.json` before the vocab file fallback.

**Current code (after PR #19 merges):**
```csharp
internal static Tokenizer LoadTokenizer(string path)
{
    if (Directory.Exists(path))
        return LoadFromDirectory(path);

    var fileName = Path.GetFileName(path).ToLowerInvariant();
    if (fileName == "tokenizer_config.json")
        return LoadFromConfig(path);

    return LoadFromVocabFile(path);
}
```

**Change to:**
```csharp
internal static Tokenizer LoadTokenizer(string path)
{
    if (Directory.Exists(path))
        return LoadFromDirectory(path);

    var fileName = Path.GetFileName(path).ToLowerInvariant();
    if (fileName == "tokenizer_config.json")
        return LoadFromConfig(path);

    if (fileName == "tokenizer.json")
        return LoadFromHuggingFaceTokenizerJson(path);

    return LoadFromVocabFile(path);
}
```

### Step 2: Handle `tokenizer.json` in directory scan in `LoadFromDirectory()` (~line 177)

Add `tokenizer.json` as a **last-resort fallback** in the directory scan  AFTER `tokenizer_config.json` and all known vocab files. This ensures existing behavior is preserved (if `vocab.txt` or `tokenizer.model` exists, those take precedence).

**Add this block just before the `throw new FileNotFoundException(...)` at the end of `LoadFromDirectory()`:**

```csharp
// Last resort: check for HuggingFace fast tokenizer file
var tokenizerJson = Path.Combine(directory, "tokenizer.json");
if (File.Exists(tokenizerJson))
    return LoadFromHuggingFaceTokenizerJson(tokenizerJson);
```

### Step 3: Implement `LoadFromHuggingFaceTokenizerJson()`  the new method

Add this new private static method to the class. This is the core implementation.

```csharp
/// <summary>
/// Loads a tokenizer from a HuggingFace tokenizer.json (fast tokenizer format).
/// Supports BPE and WordPiece model types. Unigram is not supported  use the
/// .model file instead.
/// </summary>
private static Tokenizer LoadFromHuggingFaceTokenizerJson(string tokenizerJsonPath)
{
    var json = File.ReadAllText(tokenizerJsonPath);
    using var doc = JsonDocument.Parse(json);
    var root = doc.RootElement;

    if (!root.TryGetProperty("model", out var model))
        throw new InvalidOperationException(
            $"tokenizer.json at '{tokenizerJsonPath}' does not contain a 'model' property.");

    var modelType = model.TryGetProperty("type", out var typeProp)
        ? typeProp.GetString() ?? ""
        : "";

    return modelType switch
    {
        "BPE" => LoadBpeFromTokenizerJson(model),
        "WordPiece" => LoadWordPieceFromTokenizerJson(model),
        "Unigram" => throw new NotSupportedException(
            $"Unigram model type in tokenizer.json is not supported. " +
            $"Point TokenizerPath at the directory containing the .model file instead, " +
            $"or use the Tokenizer property to provide a pre-constructed instance."),
        _ => throw new NotSupportedException(
            $"Unsupported model type '{modelType}' in tokenizer.json at '{tokenizerJsonPath}'. " +
            $"Supported types: BPE, WordPiece.")
    };
}
```

### Step 4: Implement `LoadBpeFromTokenizerJson()`  extract BPE vocab and merges from JSON

The `model` section for BPE looks like:
```json
{
  "type": "BPE",
  "vocab": { "<s>": 0, "<pad>": 1, "Ġthe": 5, ... },
  "merges": [ "Ġ t", "Ġ a", "h e", "i n", ... ]
}
```

We need to convert these to in-memory streams that `BpeTokenizer.Create(vocabStream, mergesStream)` can consume.

- `vocabStream` expects a JSON file with the format `{ "token": id, ... }`  which is exactly what `model.vocab` already is.
- `mergesStream` expects a text file with one merge per line  which is what `model.merges` contains.

```csharp
private static Tokenizer LoadBpeFromTokenizerJson(JsonElement model)
{
    if (!model.TryGetProperty("vocab", out var vocab))
        throw new InvalidOperationException(
            "BPE model in tokenizer.json does not contain a 'vocab' property.");

    // Write vocab as JSON to a MemoryStream (BpeTokenizer.Create expects a JSON stream)
    using var vocabStream = new MemoryStream();
    using (var writer = new Utf8JsonWriter(vocabStream))
    {
        vocab.WriteTo(writer);
    }
    vocabStream.Position = 0;

    // Write merges as newline-separated text to a MemoryStream
    MemoryStream? mergesStream = null;
    if (model.TryGetProperty("merges", out var merges) && merges.ValueKind == JsonValueKind.Array)
    {
        mergesStream = new MemoryStream();
        using var sw = new StreamWriter(mergesStream, leaveOpen: true);
        foreach (var merge in merges.EnumerateArray())
        {
            sw.WriteLine(merge.GetString());
        }
        sw.Flush();
        mergesStream.Position = 0;
    }

    try
    {
        return BpeTokenizer.Create(vocabStream, mergesStream);
    }
    finally
    {
        mergesStream?.Dispose();
    }
}
```

### Step 5: Implement `LoadWordPieceFromTokenizerJson()`  extract WordPiece vocab from JSON

The `model` section for WordPiece looks like:
```json
{
  "type": "WordPiece",
  "vocab": { "[PAD]": 0, "[UNK]": 100, "the": 1996, "##ing": 2075, ... }
}
```

We need to convert this to a `vocab.txt` format (one token per line, ordered by ID).

```csharp
private static Tokenizer LoadWordPieceFromTokenizerJson(JsonElement model)
{
    if (!model.TryGetProperty("vocab", out var vocab))
        throw new InvalidOperationException(
            "WordPiece model in tokenizer.json does not contain a 'vocab' property.");

    // Convert vocab dict to sorted list of (token, id) and write as vocab.txt format
    var entries = new SortedList<int, string>();
    foreach (var prop in vocab.EnumerateObject())
    {
        entries[prop.Value.GetInt32()] = prop.Name;
    }

    using var vocabStream = new MemoryStream();
    using (var sw = new StreamWriter(vocabStream, leaveOpen: true))
    {
        foreach (var entry in entries)
        {
            sw.WriteLine(entry.Value);
        }
    }
    vocabStream.Position = 0;

    return BertTokenizer.Create(vocabStream);
}
```

### Step 6: Update doc comments on `TextTokenizerOptions.TokenizerPath` (~line 25)

Update the XML doc to mention `tokenizer.json` support:

```csharp
/// <summary>
/// Path to tokenizer artifacts. Can be:
/// <list type="bullet">
///   <item>A directory containing <c>tokenizer_config.json</c>  auto-detects tokenizer type from HuggingFace config</item>
///   <item>A <c>tokenizer_config.json</c> file directly  reads <c>tokenizer_class</c> and loads sibling files</item>
///   <item>A <c>tokenizer.json</c> file (HuggingFace fast tokenizer)  supports BPE and WordPiece models</item>
///   <item>A vocab file: <c>.txt</c> (BERT/WordPiece), <c>.model</c> (SentencePiece)</item>
///   <item>A directory containing <c>tokenizer.json</c> (used as last-resort fallback)</item>
/// </list>
/// Used only when <see cref="Tokenizer"/> is not set.
/// </summary>
```

### Step 7: Update error message in `LoadFromDirectory()` 

Update the `FileNotFoundException` message to include `tokenizer.json`:

```
$"Expected one of: tokenizer_config.json, vocab.txt, tokenizer.model, sentencepiece.bpe.model, spm.model, tokenizer.json."
```

---

### Test 1: BPE tokenizer.json (direct file path)
1. Download `tokenizer.json` from [SamLowe/roberta-base-go_emotions](https://huggingface.co/SamLowe/roberta-base-go_emotions/resolve/main/tokenizer.json)
2. Set `TokenizerPath = "path/to/tokenizer.json"`
3. Verify tokenizer loads and produces correct token IDs for "Hello, world!"
4. Compare output with passing `TokenizerPath` at a directory containing `vocab.json` + `merges.txt` for the same model  IDs should match

### Test 2: WordPiece tokenizer.json (direct file path)
1. Download `tokenizer.json` from [bert-base-uncased](https://huggingface.co/google-bert/bert-base-uncased/resolve/main/tokenizer.json) (a WordPiece model)
2. Set `TokenizerPath = "path/to/tokenizer.json"`
3. Verify tokenizer loads and produces correct token IDs
4. Compare with `vocab.txt`-based loading  IDs should match

### Test 3: Directory containing only tokenizer.json
1. Create a directory with only `tokenizer.json` (no `tokenizer_config.json`, no `vocab.txt`)
2. Set `TokenizerPath = "path/to/directory/"`
3. Verify it falls through to `tokenizer.json` and loads correctly

### Test 4: Backward compatibility  existing files take precedence
1. Create a directory with BOTH `tokenizer_config.json` + `vocab.txt` AND `tokenizer.json`
2. Set `TokenizerPath = "path/to/directory/"`
3. Verify `tokenizer_config.json` is used (not `tokenizer.json`)  existing behavior preserved

### Test 5: Unigram rejection
1. Download `tokenizer.json` from a Unigram model (e.g., [DeBERTa](https://huggingface.co/MoritzLaurer/DeBERTa-v3-base-mnli-fever-anli/resolve/main/tokenizer.json))
2. Set `TokenizerPath = "path/to/tokenizer.json"`
3. Verify it throws `NotSupportedException` with a helpful message

### Test 6: Invalid/malformed tokenizer.json
1. Pass a JSON file missing the `model` property
2. Verify it throws `InvalidOperationException` with a clear message

---

## Edge Cases to Handle

- **`model.merges` is absent or empty**: Some BPE models may not have merges. `BpeTokenizer.Create` accepts a null merges stream  this is already handled in the code above.
- **`model.vocab` values may not be contiguous**: The WordPiece conversion sorts by ID, so gaps are fine  `BertTokenizer.Create` reads line-by-line.
- **Large files**: `tokenizer.json` can be 2-17 MB. `File.ReadAllText` + `JsonDocument.Parse` handles this fine for one-time loading.
- **BOM or encoding**: HuggingFace files are UTF-8 without BOM. `StreamWriter` defaults to UTF-8, which is correct.

---

## What NOT to Do

- **Do NOT add new NuGet dependencies**  this uses only `System.Text.Json` (already imported) and `Microsoft.ML.Tokenizers` (already a dependency).
- **Do NOT attempt to support Unigram models from tokenizer.json**  there's no way to construct a `LlamaTokenizer` from JSON vocab+scores (it requires a protobuf `.model` file). Unigram models always have the `.model` file available.
- **Do NOT change the priority order in `LoadFromDirectory()`**  `tokenizer_config.json` and known vocab files MUST take precedence. `tokenizer.json` is a last-resort fallback in directory mode.

---

## Acceptance Criteria

- [ ] `TokenizerPath = "path/to/tokenizer.json"` loads BPE tokenizers correctly
- [ ] `TokenizerPath = "path/to/tokenizer.json"` loads WordPiece tokenizers correctly
- [ ] `TokenizerPath = "path/to/tokenizer.json"` throws clear error for Unigram models
- [ ] Directory containing only `tokenizer.json` is auto-detected as last-resort fallback
- [ ] Backward compatible  existing `tokenizer_config.json` / `vocab.txt` / `.model` paths still work and take precedence
- [ ] Doc comments on `TokenizerPath` updated to mention `tokenizer.json`
- [ ] Build succeeds with no warnings

### PR DESCRIPTION
**Pull request by luisquintanilla**
## Summary

Adds support for loading tokenizers from HuggingFace's `tokenizer.json` fast tokenizer format  the single-file format that contains the complete tokenizer pipeline (vocab, merges, normalizer, pre/post-processing) present on virtually every model on the Hub.

Closes #21
Related: #18

### `LoadTokenizer()` (line ~175)
- Added `tokenizer.json` as a recognized file name (between `tokenizer_config.json` and vocab file fallback)

### `LoadFromDirectory()` (line ~204)
- Added `tokenizer.json` as **last-resort fallback**  only used when no `tokenizer_config.json`, `vocab.txt`, or `.model` files exist in the directory. This preserves all existing behavior.

### New methods

| Method | Purpose |
|--------|---------|
| `LoadFromHuggingFaceTokenizerJson()` | Parses JSON, reads `model.type`, dispatches to BPE/WordPiece. Throws `NotSupportedException` for Unigram with guidance. |
| `LoadBpeFromTokenizerJson()` | Extracts `model.vocab` (JSON dict, same format as vocab.json) + `model.merges` (JSON array), creates in-memory streams, calls `BpeTokenizer.Create()` |
| `LoadWordPieceFromTokenizerJson()` | Extracts `model.vocab`, sorts by ID, writes vocab.txt format, calls `BertTokenizer.Create()` |

### Doc comments & error messages
- Updated `TokenizerPath` XML doc to list `tokenizer.json` as a supported format
- Updated `LoadTokenizer()` XML doc
- Updated `LoadFromDirectory()` error message to include `tokenizer.json`

## Design decisions

1. **Unigram  throw, don't attempt**: There's no way to construct a `LlamaTokenizer` from JSON vocab+scores (it requires protobuf `.model`). Unigram models always ship `.model` files alongside `tokenizer.json`, so users should point at the directory instead.

2. **Last-resort in directory mode**: `tokenizer.json` is checked only after all other known files. This ensures zero behavioral change for existing users.

3. **No new dependencies**: Uses only `System.Text.Json` (already imported) and `Microsoft.ML.Tokenizers` (already a dependency).

## Testing suggestions

| # | Scenario | Model to test with |
|---|----------|-------------------|
| 1 | BPE tokenizer.json (direct file) | [SamLowe/roberta-base-go_emotions](https://huggingface.co/SamLowe/roberta-base-go_emotions/resolve/main/tokenizer.json) |
| 2 | WordPiece tokenizer.json (direct file) | [bert-base-uncased](https://huggingface.co/google-bert/bert-base-uncased/resolve/main/tokenizer.json) |
| 3 | Directory with only tokenizer.json | Any BPE model |
| 4 | Backward compat (dir with vocab.txt + tokenizer.json) | Any BERT model  should use vocab.txt |
| 5 | Unigram rejection | [DeBERTa](https://huggingface.co/MoritzLaurer/DeBERTa-v3-base-mnli-fever-anli/resolve/main/tokenizer.json)  throws NotSupportedException |
| 6 | Malformed tokenizer.json | Missing `model` property  throws InvalidOperationException |

### CODE REVIEW

File: src/MLNet.TextInference.Onnx/TextTokenizerEstimator.cs [hunk 1]
```diff
@@ -286,6 +298,92 @@ private static Tokenizer LoadBpeFromDirectory(string directory)
...
+        if (!root.TryGetProperty("model", out var model))
+            throw new InvalidOperationException(
+                $"tokenizer.json at '{path}' has no 'model' property. " +
+                $"Ensure this is a valid HuggingFace fast tokenizer file.");
+
+        if (!model.TryGetProperty("type", out var typeElement))
+            throw new InvalidOperationException(
+                $"tokenizer.json at '{path}' has no 'model.type' property.");
+
+        var modelType = typeElement.GetString();
```
[the remark this answers is not in the document]
Comment (luisquintanilla) [reply src/MLNet.TextInference.Onnx/TextTokenizerEstimator.cs]: **Accepted.** Null model.type would fall into the catch-all with a confusing empty-string error. Fixed in a861166 with null-coalescing throw.
